### DISCUSSION TITLE
Can I create a headless app with an OpenGL context for my unit tests?

### DISCUSSION THREAD

**Discussion by nateglasser** [AvaloniaUI/Avalonia#15761] [Q&A] [answered]
I have a set of SkSl shaders for manipulating large images (running on the experimental SkiaSharp 3 support). I want to add tests to ensure the shader output is consistent with a reference software implementation.

I set up my XUnit tests as described [here](https://docs.avaloniaui.net/docs/concepts/headless/headless-xunit). That lets me use Skia's software path, which gets me part of the way there, but what I really need for my tests is a hardware-accelerated GRContext. In my app I can get a shared OpenGL context as described in #13261 and use it for `GRContext.CreateGl`, but I can't figure out anything similar for the testing framework. I tried to set up the AppBuilder to use a specific platform, e.g. with UseWin32(), but I still get a headless platform without an OpenGL context.

I can always go through the ceremony of manually creating my own OpenGL context, but it would be nice to do it using Avalonia for the current platform. Is there a recommended way to create an OpenGL context for the testing frameworks?

**Comment by maxkatz6** [comment, marked as answer]
Currently it's not possible, and headless platform doesn't use any hardware acceleration.

You can try testing with Appium, basically running a real app with a real window, but whole setup is pretty difficult - https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/Testing/TestableApp.Appium

**Reply by nateglasser** [reply]
Thanks for confirming, I will check out Appium and decide if it's worth it.
